enshrine-esdt-safe/src/to_sovereign/create_tx.rs

```diff
@@ -231,92 +60,73 @@ pub trait CreateTxModule:
-                let sov_token_id = self
-                    .multiversx_to_sovereign_token_id(&payment.token_identifier)
-                    .get();
-
-                if !sov_token_id.is_valid_esdt_identifier() {
-                    event_payments.push(MultiValue3((
-                        payment.token_identifier,
-                        payment.token_nonce,
-                        current_token_data.clone(),
-                    )));
-
-                    continue;
-                }
-
-                let sov_token_nonce = self.remove_sovereign_token(payment, &sov_token_id);
-
-                event_payments.push(MultiValue3((
-                    sov_token_id,
-                    sov_token_nonce,
-                    current_token_data.clone(),
-                )));
-            }
+                    current_token_data,
+                )
+                    .into(),
+            );
         }
 
-        let caller = self.blockchain().get_caller();
+        let option_transfer_data = TransferData::from_optional_value(opt_transfer_data);
 
-        self.match_fee_payment(total_tokens_for_fees, &fees_payment, &opt_transfer_data);
+        if let Some(transfer_data) = option_transfer_data.as_ref() {
+            self.require_gas_limit_under_limit(transfer_data.gas_limit);
+            self.require_endpoint_not_banned(&transfer_data.function);
+        }
+        self.match_fee_payment(total_tokens_for_fees, &fees_payment, &option_transfer_data);
 
         // refund refundable_tokens
-        for payment in &refundable_payments {
-            self.send().direct_non_zero_esdt_payment(&caller, &payment);
-        }
+        let caller = self.blockchain().get_caller();
+        self.refund_tokens(&caller, &refundable_payments);
 
         let tx_nonce = self.get_and_save_next_tx_id();
-
         self.deposit_event(
             &to,
             &event_payments,
-            OperationData {
-                op_nonce: tx_nonce,
-                op_sender: caller,
-                opt_transfer_data,
-            },
+            OperationData::new(tx_nonce, caller, option_transfer_data),
         );
     }
 
-    fn remove_sovereign_token(
+    fn check_and_extract_fee(
         &self,
-        payment: EsdtTokenPayment<Self::Api>,
-        sov_token_id: &TokenIdentifier<Self::Api>,
-    ) -> u64 {
-        self.send().esdt_local_burn(
-            &payment.token_identifier,
-            payment.token_nonce,
-            &payment.amount,
-        );
+    ) -> MultiValue2<OptionalValue<EsdtTokenPayment>, ManagedVec<EsdtTokenPayment>> {
+        let mut payments = self.call_value().all_esdt_transfers().clone_value();
 
-        let mut sov_token_nonce = 0;
+        require!(!payments.is_empty(), "Nothing to transfer");
+        require!(payments.len() <= MAX_TRANSFERS_PER_TX, "Too many tokens");
+
+        let fee_market_address = self.fee_market_address().get();
+        let fee_enabled = self.external_fee_enabled(fee_market_address).get();
```

Comment:
If `fee_market_address` is not set, will this return `panic`? You have to make sure the address exists in storage.